bittensor/_dataset/dataset_impl.py

```diff
@@ -542,137 +537,494 @@ def set_data_size(self, batch_size, block_size):
             block_size(int, required):
                 The block_size of data that should be produced by dataloader. 
         """
+
+        assert self.run_generator == False, 'Please ensure that run_generator is false'
+
         def check_valid(size):
             r""" Check if the size is a valid positive intiget, if not, return False.
             """
             if size <= 0 or (not isinstance(size, int)):
                 return False
             else:
                 return True
-
-        old_batch_size = self.batch_size
-        old_block_size = self.block_size
         
         if check_valid(batch_size):
             self.batch_size = batch_size
         
         if check_valid(block_size):
             self.block_size = block_size
+            # rebuild the cat params list
+            self.sample_cat_params_list = self.build_sample_cat_params(all_text_file_metas=self.all_text_file_metas, block_size=self.block_size)
 
         # empty the queue
-        while not self.data_queue.queue.empty():
-            self.data_queue.queue.get()
-
+        if self.run_generator:
+            while not self.data_queue.empty():
+                self.data_queue.get()
+        else:
+            # empty sample cache
+            self.set_cache(cache_size=self.cache_size)
+            self.sample_cache = []
         # empty the dataset_iterator with the old sizing
-        self.__infinite_dataset_iterator = iter([])
+        self.__infinite_dataset_iterator = None
+
+    def set_cache(self, cache_size=10) -> list:
+        '''
+        Set the cache and ensure it is valid.
+        '''
+        self.sample_cache = [] # Cache list for raw text.
+        self.cache_size = min(cache_size, len(self) )  # The maximum size of the cache (FIFO).
+        self.cache_size = (self.cache_size // self.batch_size) * self.batch_size
+        return self.sample_cache
 
-        logger.success(f"Updated data size: batch_size: {old_batch_size} --> {self.batch_size}, block_size: {old_block_size} --> {self.block_size}")
+    def resolve_sample_cache(self)-> List[str]:
+        '''
+        Checks the sample cache, and builds it if it is empty
+
+        Returns:
+            self.sample_cache (List[str]): 
+                The sample cache.
+        '''
+        if len(self.sample_cache) == 0:
+            sample_cat_params_list = random.sample(self.sample_cat_params_list, self.cache_size)
+            cache_sample_tasks = [self.cat(**sample_cat_params) for sample_cat_params in sample_cat_params_list]
+            self.sample_cache = [sample_bytes for sample_bytes in self.async_run(asyncio.gather(*cache_sample_tasks))]
 
-    def dataloader(self, epoch_length = 100):
-        r""" Creates a torch dataloader out of a subclass of this class.
+
+        return self.sample_cache
+
+    def __getitem__(self, idx: Optional[int]= None, filler_token:str='FILLER_TEXT') -> Union[List[str], torch.tensor]:
```

Comment:
probably should remove `filler_token` since we default to `self.pad_token` to avoid ambiguity.